lib/src/secure_config.rs

```diff
@@ -0,0 +1,598 @@
+        Ok((path, metadata))
+    }
+
+    /// Validates that the metadata path matches the repo path.
+    /// If there's a mismatch, takes appropriate action.
+    /// Returns the updated config dir and metadata.
+    fn handle_metadata_path(
+        &self,
+        rng: &mut ChaCha20Rng,
+        root_config_dir: &Path,
+        config_dir: PathBuf,
+        mut metadata: ConfigMetadata,
+    ) -> Result<LoadedSecureConfig, SecureConfigError> {
+        let encoded = path_to_bytes(&self.repo_dir).ok();
+        let got = metadata
+            .path
+            .as_deref()
+            .map(|p| path_from_bytes(p))
+            .transpose()?;
+
+        if got == encoded.is_some().then_some(self.repo_dir.as_path()) {
+            return Ok(LoadedSecureConfig {
+                config_file: Some(config_dir.join(CONFIG_FILE)),
+                metadata,
+                warnings: vec![],
+            });
+        }
+        let got = match got {
+            Some(d) if d.is_dir() => d.to_path_buf(),
+            _ => {
+                // The old repo does not exist. Assume the user moved it.
+                metadata.path = encoded.map(|b| b.to_vec());
+                update_metadata(&config_dir, &metadata)?;
+                return Ok(LoadedSecureConfig {
+                    config_file: Some(config_dir.join(CONFIG_FILE)),
+                    metadata,
+                    warnings: vec![],
+                });
+            }
+        };
+        // We attempt to create a temporary file in the new repo.
+        // If it fails, we have readonly access to a repo, so we do nothing.
+        if let Ok(tmp) = NamedTempFile::new_in(&self.repo_dir) {
+            // If we write to the new repo and it shows up in the old one,
+            // we can skip this step, since it's not a copy.
+            if !got.join(tmp.path().file_name().unwrap()).exists() {
+                // We now assume the repo was copied. Since the repo was copied,
+                // the config should be copied too, rather than sharing the
+                // config with what it copied from.
+                let old_config_path = config_dir.join(CONFIG_FILE);
+                metadata.path = encoded.map(|b| b.to_vec());
+                let old_config_content = fs::read(&old_config_path).context(&old_config_path)?;
+                let config_path = self.generate_config(
+                    root_config_dir,
+                    &generate_config_id(rng),
+                    Some(&old_config_content),
+                    &metadata,
+                )?;
+                return Ok(LoadedSecureConfig {
+                    config_file: Some(config_path.clone()),
+                    metadata,
+                    warnings: vec![format!(
+                        "Your repo appears to have been copied from {got:?} to {:?}. The \
+                         corresponding repo config file has also been copied.",
+                        &self.repo_dir
```

Comment:
Use `.display()` instead of debug formatting. Windows directory separators would look ugly with `{:?}`.